requirements: updated Interrupts sections

Timotheous opened the pull request:
updated system_requirements Interrupts section
              updated software_requirements Interrupts section

simhein commented on docs/system_requirements/index.sdoc:
```diff
@@ -97,12 +97,20 @@ UID: ZEP-39
-As a Zephyr RTOS user I want interrupts to be handled synchronously in response to a hardware or software interrupt request with a minimum latency, preempting threads and, as far as the hardware allows, lower priority interrupt service routines.
+As the Zephyr RTOS user I want the Kernel to mask the HAL interfaces so I can exclusively use Zephyr Kernel calls to configure and enable interrupt hardware and install interrupt software handlers.
+<<<
+REVIEW_COMMENT: >>>
+Tim: The user story depends on understanding the role of the HAL and Architecture Interfaces.  
+This understanding will also impact the Software Requirements. Should we add:
+'Given that interrupt configuration and behavior is hardware dependent, handled by the 
+platform specific HAL, and accessed by a predefined set of Architecture interfaces.' OR
+
+Do interrupts belong under Architecture.
```
I don't think adding the interrupts under architecture is needed there should be a requirement under the Architecture interfaces which can be connected to this requirements.

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+COMPONENT: Interrupts
+TITLE: Uninstalling dynamic IRQ service routines (ISR).
+STATEMENT: >>>
+Zephyr RTOS shall provide a mechanism to uninstall a dynamic ISR and reset the IRQ hardware. 
```
Does this really rests the IRQ hardware?
In my opinion the hardware reset should be removed here because as far as I understand the functionality it only disconnects the dynamic ISR is removed from the association with the irq itself.

Timotheous replied:
Right!  I'll fix that.

Timotheous replied:
Done

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+COMPONENT: Interrupts
+TITLE: ISR Context status
+STATEMENT: >>>
+Zephyr RTOS shall provide mechanisms that returns the execution context of In-ISR or Not In-ISR.
```
```suggestion
Zephyr RTOS shall provide mechanisms that returns the execution context, where the context is In-ISR or Not In-ISR.
```

Timotheous replied:
Done

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+COMPONENT: Interrupts
+TITLE: Kernel Initialization status
+STATEMENT: >>>
+Zephyr RTOS shall provide mechanisms that returns the Kernel execution context 
+of Pre-Kernel, or In-Kernel.
```
```suggestion
Zephyr RTOS shall provide mechanisms that returns the Kernel execution context, where the context is Pre-Kernel or In-Kernel.
```

Timotheous replied:
Done

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+UID: ZEP-13013
 STATUS: Draft
 TYPE: Functional
 COMPONENT: Interrupts
-TITLE: Disabling interrupts
+TITLE: Thread Preemptable status
 STATEMENT: >>>
-The Zephyr RTOS shall provide mechanisms to disable interrupts.
+Zephyr RTOS shall provide mechanisms that indicates if the 
+current execution context is a preemptable thread or not.
 <<<
 USER_STORY: >>>
-As a Zephyr RTOS user I want to be able to disable each individual and a global interrupt according to my applications needs during program execution.
+As the developer of functions that may run in either a preemptable thread or other context, 
+I need to know the current execution context to enable condition behavior.
+<<<
+REVIEW_COMMENT: >>>
+Tim: Not relevant to the Interrupts requirements.  
+But I found it nearby and added it so it does not get lost.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Also this might need to be updated in the documentation because the requirement belongs to https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#c.k_is_preempt_thread in the documentation under the interrupt section

Timotheous replied:
Move it up to "Threads" 

This call would be used by code that could run in a thread after the scheduler starts, or prior to the scheduler starting as part of the initialization process..

Timotheous replied:
Removed for inclusion in a (near) future commit to the threads software requirements

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+TITLE: Multi-level interrupts
+STATEMENT: >>>
+The Zephyr RTOS shall support multi-level preemptive interrupt priorities, when supported by hardware.
+Note: detailed analysis to demonstrate non interference will be needed here.
```
Is this note needed? I think we should remove such notes and this should be proven by a test. Or did I get the note here wrong?

Timotheous replied:
I think you got the note correct.  I'll remove it.

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
-The Zephyr RTOS shall provide support a service routine for handling interrupts (ISR).
+Zephyr RTOS shall provide a mechanism to initialize a static IRQ service routine (ISR),
+providing all parameters needed to configure the hardware and software.
+The static IRQ shall be initially disabled.
```
It's a reasonable compound requirement.  The requirement is about initializing the IRQ, and the state after initialization is an integral part. I could make the statement more complex so it looks more like one requirement.

All IRQs are initially disabled, but that may not always be true. 
So breaking it out is not unreasonable either.

Easily done. 
If someone else thinks two requirements are better I will do it.

Timotheous replied:
Done

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
 STATUS: Draft
 TYPE: Functional
 COMPONENT: Interrupts
-TITLE: Associating application code with interrupts
+TITLE: Direct ISR, Platform Specific helper macros.
 STATEMENT: >>>
-The Zephyr RTOS shall provide an interface for associating application code with specific interrupts. (CLARIFY: Can it be a deferred procedure call at interrupt context? Would be different requirement)
+Zephyr RTOS shall provide macros supporting development of direct ISRs.
+ * Minimal header instructions
+ * Minimal footer instructions
+ * Low Power termination
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
I was not sure where to put it.  System level works for me.

Timotheous replied:
Moved to System Requirements

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+  VALUE: ZEP-39
+
+[REQUIREMENT]
+UID: ZEP-13012
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Interrupts
+TITLE: Kernel Initialization status
+STATEMENT: >>>
+Zephyr RTOS shall provide mechanisms that returns the Kernel execution context 
+of Pre-Kernel, or In-Kernel.
 <<<
 USER_STORY: >>>
-As a Zephyr RTOS user I want to be able to enable each individual and a global interrupt according to my applications needs during program execution.
+As the developer of functions that may run either before or after Kernel initialization, 
+I need to know the current Kernel Initialization Status to enable condition behavior.
 <<<
+REVIEW_COMMENT: >>>
+Tim: Not relevant to the Interrupts requirements.  
+But I found it nearby and added it so it does not get lost.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Move it up to "Threads" 

This call would be used by code that could run in a thread after the scheduler starts, or prior to the scheduler starting as part of the initialization process..

Timotheous replied:
Removed for inclusion in a (near) future commit to the threads software requirements

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+Zephyr RTOS shall provide a mechanism to initialize an direct IRQ handler,
+providing all parameters needed to configure the hardware and software.
+
+The direct IRQ shall be initially disabled.
```
Done

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
-The Zephyr RTOS shall provide mechanisms to enable interrupts.
+Zephyr RTOS shall provide a mechanism to initialize a dynamic IRQ service routine, 
+providing all parameters needed to configure the hardware and software.
+The direct IRQ shall be initially disabled.
```
Done

Timotheous commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -6,57 +6,228 @@ REQ_PREFIX: ZEP-
+The Zephyr RTOS shall support multi-level preemptive interrupt priorities, when supported by hardware.
+Note: detailed analysis to demonstrate non interference will be needed here.
+<<<
+REVIEW_COMMENT: >>>
+Tim: This functionality has found in the RTOS, or in any of the Architecuture APIs.
+It is dependant on the platform,the HAL layer, and the implementation of the ISR.  
+Perhaps I missed something.
+It is described in the Interrupts documentation.  So I left it in.
```
Removed them all.  Saved a few to turn into proper review comments

simhein commented on .gitignore:
```diff
@@ -11,3 +11,7 @@ Copy of Zephyr LTS Requirements - 20220301.xlsx
 output/
 output.zip
 
+# My Misc
+ReqDocsHtml.bat
+ReqDocsHtml2Pdf.bat
+docs-Copy/
```
The personal gitignore files should be removed from the PR

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -1,62 +1,204 @@
+TITLE: Installing low-power and low-latency direct IRQ service routines (ISR).
 STATEMENT: >>>
-The Zephyr RTOS shall provide an interface for associating application code with specific interrupts. (CLARIFY: Can it be a deferred procedure call at interrupt context? Would be different requirement)
+Zephyr RTOS shall provide a mechanism to initialize a direct IRQ handler,
+providing all parameters needed to configure the hardware and software.
```
What about the low power and low latency which was mentioned in the Title?

Timotheous replied:
A Direct IRQ Handler allows customizations to the IRQ handler via macros that allow developer defined abbreviated setup and teardown code specific to the use-case being supported.  Also allows developer to enable, or not, exiting of CPU LowPowerMode.  See ZEP-39-1, Direct ISR, Platform Specific helpers.

The "standard" IRQ installer implementation includes general purpose setup and teardown code, which is necessarily much slower. And always exits LPM. 

All this is designed to enable low power and fast interrupts without modifying the HAL.  But still requires someone at in the organization to write the setup, teardown, and lpm code fragments that are correct for the platform.  The alternative is to modify the HAL.  There are trade-offs.

Timotheous replied:
Will be cleaner if Title is changed to:
Installing Direct IRQ serivce routines (ISR).

simhein replied:
> Will be cleaner and more correct if Title is changed to: Installing Direct IRQ service routines (ISR).

That is a good point I think that change clarify this topic good

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -1,62 +1,204 @@
+TITLE: Multi-level interrupts
+STATEMENT: >>>
+The Zephyr RTOS shall support multi-level preemptive interrupt priorities, when supported by hardware.
+Note: detailed analysis to demonstrate non interference will be needed here.
```
Note needs still to be removed.

Timotheous replied:
Done

simhein commented:
@Timotheous Sorry for my really late re-review on that. It looks very good now. Please rebase this PR aswell

simhein reviewed (approved):
LGTM now

simhein commented on .gitignore:
```diff
@@ -11,3 +11,7 @@ Copy of Zephyr LTS Requirements - 20220301.xlsx
+# My Misc
+ReqDocsHtml.bat
+ReqDocsHtml2Pdf.bat
+docs-Copy/
```
The private gitignore files still needs to be removed

Timotheous replied:
done

simhein commented on docs/system_requirements/index.sdoc:
```diff
@@ -91,14 +91,36 @@ UID: ZEP-SYRS-7
 <<<
 
+[REQUIREMENT]
+UID: ZEP-39-1
```
This Req UID don't align with the System requirements UIDs Layout

Timotheous replied:
Not sure how to handle this.  It is a system requirement and links to ZEP-SYRS-7 but no others.
Do I use the next available ZEP-SYRS-#, or do I use -1?

Either works for me, what would you prefer.
I used ZEP-SYRS-7-1 to make strictdoc run.

simhein replied:
The easiest way is to delete the line with the UID locally and than execute the strict doc command `strictdoc manage auto-uid .` Than the next free SYRS UID is selected or as you described it using the next available TEP-SYRS-#

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

Timotheous replied:
updated all the ZEP-39 tags to ZEP-SYRS-7

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/software_requirements/interrupts.sdoc:
```diff
@@ -10,53 +10,194 @@ UID: ZEP-SRS-7-1
 <<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Link need to be adapted

simhein commented on docs/system_requirements/index.sdoc:
```diff
@@ -91,14 +91,36 @@ UID: ZEP-SYRS-7
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-39
```
Shall this link to itself?

simhein commented:
> LGTM now

@Timotheous Sorry I was a little bit fast with the approval. There is a CI error due to the not found Requirement UIDs in the linkage This needs to be fixed to make the CI pass.

Timotheous commented:
My fault.  I failed to run strictdoc and build the HTML.

Updated the offending links.
Strictdoc now runs locally.
Squashed and force pushed to PR.

Seems correct now. Except for one question above.

Timotheous commented:
Not sure of the issue with .gitignore.
I removed all the lines that I added.

simhein commented:
> Not sure of the issue with .gitignore. I removed all the lines that I added.

True that was just a odd view from the github review diff view. You can ignore that comment

Timotheous commented:
Change made.  Should be good now.

simhein reviewed (approved):
Looking good now thanks!
